readme.md
# jQuery.DeferredPipeline

manage async processes.

## Demos

* [imgs' preload](http://takazudo.github.io/jQuery.DeferredPipeline/demos/1/)

## Tests

* [all](http://takazudo.github.io/jQuery.DeferredPipeline/tests/mocha/common/)

## What this does

Basically, if you call 4 async processes, It's like this.

![](mocks/noPipe.png)

With jQuery.DeferredPipeline, you can queue functions like this.

![](mocks/pipeSize1.png)

```javascript
var pipeline = new $.DeferredPipeline({ pipeSize: 1 });

// queue function like this 4 times
pipeline.add(function() {
  var d = $.Deferred();
  doAsync(function() {
    d.resolve();
  });
  return d.promise();
});

pipeline.run();
```

This is just a queue. You can run specified counts' async processes at the same time like this.

![](mocks/pipeSize2.png)

```javascript
// 2 size pipeline
var pipeline = new $.DeferredPipeline({ pipeSize: 2 });

// queue function like this 8 times
pipeline.add(function() {
  var d = $.Deferred();
  doAsync(function() {
    d.resolve();
  });
  return d.promise();
});

pipeline.run();
```

## Why you need this

### Images' preload

You can reduce the CPU and network traffic costs by piping.

### Ajax requests (especially JSONP)

Multiple JSONP requests may get troubled sometimes. Because it is just a hacky JavaScript execution, so we need to keep the order of the JSONP requests.

## Usage

### Basic

like the explanation above, do

```javascript
var pipeline = new $.DeferredPipeline({ pipeSize: 2 }); // specifiy pipeSize as option

// queue functions like below
pipeline.add(function() {
  var d = $.Deferred();
  doAsync(function() {
    d.resolve();
  });
  return d.promise();
});

pipeline.run();
```

### done, fail, complete

The second argument is for the option.

```javascript
var pipeline = new $.DeferredPipeline({ pipeSize: 2 });

pipeline.add(function() {
  var d = $.Deferred();
  doAsync(function() {
    d.resolve('someValue'); // or .reject()
  });
  return d.promise();
}, {
  done: function(doneArgs) {
    // called when main function's returned promise was resolved.
    // this function can receive passed arguments as `doneArgs`
  },
  fail: function(failArgs, data) {
    // called when main function's returned promise was rejected.
    // this function can receive passed arguments as `failArgs`
    // if the main function was aborted, `true` will be set to `data.aborted`.
  },
  complete: function(doneOrFailArgs, data) {
    // called after `done` or `fail` was called.
    // this function can receive passed arguments as `doneOrFailArgs`
    // if the main function was aborted, `true` will be set to `data.aborted`.
    // if `done`, `true` will be set to `data.successed`.
    // if `fail`, `false` will be set to `data.successed`.
  }
});

pipeline.run();
```

### Stop

#### stopAll

```javascript
var pipeline = new $.DeferredPipeline({ pipeSize: 2 });

pipeline.add(...);
pipeline.add(...);
pipeline.add(...);

pipeline.run();

// then, later...

pipeline.stopAll(); // stops all queued items.
```

Then, all pending items will be not called, and about already called items, you can get the aborted stats like this.

```javascript
pipeline.add(function(stats) {
  var d = $.Deferred();
  doAsync(function() {
    if(stats.aborted) { // if pipeline was stopped, `stats.aborted` will be `true`
      d.reject();
    } else {
      d.resolve();
    }
  });
  return d.promise();
});
```

### stopAllWithoutTheLast

`stopAllWithoutTheLast` does almost same stuff as `stopAll`, but this calls the last queued item this time. This may useful for queued JSONP requests.

### Events

![](mocks/events.png)

```javascript
var pipeline = new $.DeferredPipeline({ pipeSize: 2 });

pipeline.add(...);
pipeline.add(...);
pipeline.add(...);

pipeline.on('run', function() { ... }); // when run
pipeline.on('stop', function() { ... }); // when stopped
pipeline.on('itemRun', function() { ... }); // when any item was run
pipeline.on('itemSuccess', function() { ... }); // when any item was successed
pipeline.on('itemFail', function() { ... }); // when any item was failed
pipeline.on('itemComplete', function() { ... }); // when any item was completed
pipeline.on('allComplete', function() { ... }); // when all items were completed or stopped

pipeline.run();
```

## Depends

* [EveEve](https://github.com/Takazudo/EveEve)
* jQuery 1.10.2 (>=1.5.1)

## Browsers

IE6+ and other new browsers.  

## How to build

git clone, then `git submodule init`, `git submodule update`.  
Then, `grunt` to build or `grunt watch` to watch coffee file's change.

## License

Copyright (c) 2014 "Takazudo" Takeshi Takatsudo  
Licensed under the MIT license.

## Build

Use

 * [CoffeeScript][coffeescript]
 * [grunt][grunt]
 * [Mocha](http://visionmedia.github.io/mocha/) for tests
 * [LearnBoost/expect.js](https://github.com/LearnBoost/expect.js/) for tests

[coffeescript]: http://coffeescript.org "CoffeeScript"
[grunt]: http://gruntjs.com "grunt"


changelog.md
# Change log

## v0.1.0

* release


demos/1/scripts.js
(function() {

  var loadImg = function(src) {
    var defer = $.Deferred();
    var img = new Image;
    var cleanUp = function() {
      img.onload = img.onerror = null;
    };
    img.onload = function() {
      cleanUp();
      defer.resolve(img);
    };
    img.onerror = function() {
      cleanUp();
      defer.reject(img);
    };
    img.src = src;
    return defer.promise();
  };

  loadImg('../imgs/loading.gif');

  var ImgPlacer = function() {
    var html = $.trim($('#tmpl-imgPlacer').html());
    this.$el = $(html);
    this.$loading = this.$el.find('.imgPlacer__loading');
    this.$img = this.$el.find('.imgPlacer__img');
  };
  ImgPlacer.prototype = {
    putImg: function(imgEl) {
      var self = this;
      self.$img.hide();
      self.removeLoading();
      setTimeout(function() {
        self.$img.append(imgEl).fadeIn(200);
      }, 150);
    },
    toAborted: function() {
      this.$el.addClass('aborted');
    },
    showLoading: function() {
      if(this._loadingRemoved || this._loadingShown) {
        return;
      }
      this._loadingShown = true;
      this.$loading.fadeIn(200);
    },
    removeLoading: function() {
      this._loadingRemoved = true;
      var self = this;
      $.when(self.$loading.fadeOut(200)).done(function() {
        self.$loading.remove();
      });
    }
  };

  $(function() {

    $('.testBlock').each(function(i, el) {
      
      pipeSize = i + 1;

      var pipeline = new $.DeferredPipeline({ pipeSize: pipeSize });
      var $el = $(el);

      var $imgs = $('.imgs', el);

      var imgPlacers = [];

      var addPlacers = function() {
        for (var i=1; i<=20; i+=1) {
          (function(i) {
            var placer = new ImgPlacer();
            imgPlacers.push(placer);
            placer.$el.appendTo($imgs);
            pipeline.add(function() {
              var d = $.Deferred();
              loadImg('../photos/' + i + '.jpg').always(function(img) {
                placer.putImg(img);
                setTimeout(d.resolve, 150);
              });
              return d.promise();
            }, {
              fail: function(failArgs, stats) {
                placer.removeLoading();
                if(stats.aborted) {
                  placer.toAborted();
                }
              }
            });
            if(pipeline.running) {
              placer.showLoading();
            }
          }(i));
        }
      };

      var showLoadings = function() {
        for (var i=0, l=imgPlacers.length; i<l; i+=1) {
          imgPlacers[i].showLoading();
        }
      };

      $el
        .on('click', '.add', function(e) {
          addPlacers();
        })
        .on('click', '.run', function(e) {
          showLoadings();
          pipeline.run();
        })
        .on('click', '.stopAll', function(e) {
          pipeline.stopAll();
        })
        .on('click', '.stopAllWithoutTheLast', function(e) {
          pipeline.stopAllWithoutTheLast();
        })

    });
  });

}());



jquery.DeferredPipeline.js
/*! jQuery.deferredPipeline (https://github.com/Takazudo/jQuery.deferredPipeline)
 * lastupdate: 2014-03-27
 * version: 0.1.0
 * author: 'Takazudo' Takeshi Takatsudo <takazudo@gmail.com>
 * License: MIT */
(function() {
  var __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  (function($) {
    var EveEve, ns, wait;
    EveEve = window.EveEve;
    ns = {};
    ns.util = {};
    wait = function(time) {
      return $.Deferred(function(defer) {
        return setTimeout(function() {
          return defer.resolve();
        }, time);
      });
    };
    ns.MESSAGE = {
      error_no_deferred: 'Error: registered function should return a deferred'
    };
    ns.util.isPromise = function(obj) {
      if ((obj != null) && (obj.then != null) && (obj.done != null)) {
        return true;
      }
      return false;
    };
    ns.Item = (function(_super) {
      __extends(Item, _super);

      Item.defaults = {
        done: null,
        fail: null,
        complete: null
      };

      function Item(fn, options) {
        this.options = $.extend({}, ns.Item.defaults, options);
        this.started = false;
        this.running = false;
        this.stopped = false;
        this._fn = fn;
      }

      Item.prototype._attachNoDeferredMessage = function(obj) {
        return obj.msg = ns.MESSAGE.error_no_deferred;
      };

      Item.prototype._triggerSuccess = function(doneArgs) {
        var _base;
        this.stopped = true;
        if (typeof (_base = this.options).done === "function") {
          _base.done(doneArgs);
        }
        this.trigger('success', doneArgs);
        return this._triggerComplete(doneArgs, true);
      };

      Item.prototype._triggerFail = function(failArgs, aborted, noDeferred) {
        var data, _base;
        if (aborted == null) {
          aborted = false;
        }
        if (noDeferred == null) {
          noDeferred = false;
        }
        this.stopped = true;
        data = {};
        data.aborted = aborted;
        if (noDeferred) {
          this._attachNoDeferredMessage(data);
        }
        if (typeof (_base = this.options).fail === "function") {
          _base.fail(failArgs, data);
        }
        this.trigger('fail', failArgs, data);
        return this._triggerComplete(failArgs, false, aborted, noDeferred);
      };

      Item.prototype._triggerComplete = function(doneOrFailArgs, successed, aborted, noDeferred) {
        var data, _base;
        if (aborted == null) {
          aborted = false;
        }
        if (noDeferred == null) {
          noDeferred = false;
        }
        data = {
          successed: successed,
          aborted: aborted
        };
        if (noDeferred) {
          this._attachNoDeferredMessage(data);
        }
        if (typeof (_base = this.options).complete === "function") {
          _base.complete(doneOrFailArgs, data);
        }
        return this.trigger('complete', doneOrFailArgs, data);
      };

      Item.prototype.destroy = function() {
        this.off();
        return this._fn = null;
      };

      Item.prototype.stop = function() {
        var _ref;
        if (this.stopped) {
          return;
        }
        this.stopped = true;
        if ((_ref = this._completeStats) != null) {
          _ref.aborted = true;
        }
        return this._triggerFail([], true);
      };

      Item.prototype.run = function() {
        var promise;
        this._completeStats = {
          aborted: false
        };
        this.started = true;
        if (this.stopped && (!this.running)) {
          return;
        }
        this.running = true;
        promise = this._fn(this._completeStats);
        if (!ns.util.isPromise(promise)) {
          this.running = false;
          this._triggerFail([], false, false, true);
          return;
        }
        this.trigger('run');
        return promise.then((function(_this) {
          return function() {
            _this.running = false;
            if (_this.stopped) {
              return;
            }
            return _this._triggerSuccess(arguments);
          };
        })(this), (function(_this) {
          return function() {
            _this.running = false;
            return _this._triggerFail(arguments, false);
          };
        })(this));
      };

      return Item;

    })(EveEve);
    ns.Pipeline = (function(_super) {
      __extends(Pipeline, _super);

      Pipeline.defaults = {
        pipeSize: 3
      };

      function Pipeline(options) {
        this.completeCount = 0;
        this.options = $.extend({}, ns.Pipeline.defaults, options);
        this._items = [];
      }

      Pipeline.prototype._beforeStoppingItems = function() {
        this._stoppingItemsInProgress = true;
        return this.trigger('startStoppingItems');
      };

      Pipeline.prototype._afterStoppingItems = function() {
        this._stoppingItemsInProgress = false;
        this.trigger('endStoppingItems');
        return this.trigger('stop');
      };

      Pipeline.prototype._handleTheLastItemCompletion = function() {
        this.running = false;
        return (wait(0)).done((function(_this) {
          return function() {
            return _this.trigger('allComplete');
          };
        })(this));
      };

      Pipeline.prototype._handleNextRun = function() {
        if (this._stoppingItemsInProgress) {
          return this.once('endStoppingItems', (function(_this) {
            return function() {
              return _this._tryToRunNextItem();
            };
          })(this));
        } else {
          return this._tryToRunNextItem();
        }
      };

      Pipeline.prototype._findNextPendingItem = function() {
        var item, _i, _len, _ref;
        _ref = this._items;
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          item = _ref[_i];
          if (item.started === false) {
            return item;
          }
        }
        return null;
      };

      Pipeline.prototype._removeItem = function(item) {
        var current, refreshed, _i, _len, _ref;
        refreshed = [];
        _ref = this._items;
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          current = _ref[_i];
          if (current !== item) {
            refreshed.push(current);
          }
        }
        this._items = refreshed;
        return item.destroy();
      };

      Pipeline.prototype._tryToRunNextItem = function() {
        var hitLimit, pipeSize, runOne, runningCount, _results;
        pipeSize = this.options.pipeSize;
        runningCount = this.countRunningItems();
        hitLimit = false;
        runOne = (function(_this) {
          return function() {
            var next;
            next = _this._findNextPendingItem();
            if (next != null) {
              next.run();
            } else {
              hitLimit = true;
            }
            return runningCount = _this.countRunningItems();
          };
        })(this);
        _results = [];
        while ((runningCount < pipeSize) && (!hitLimit)) {
          _results.push(runOne());
        }
        return _results;
      };

      Pipeline.prototype.destroy = function() {
        var item, _i, _len, _ref;
        this.stopAll();
        this.off();
        _ref = this._items;
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          item = _ref[_i];
          item.destroy();
        }
        return this._items = null;
      };

      Pipeline.prototype.add = function(fn, options) {
        var item;
        item = new ns.Item(fn, options);
        item.on('run', (function(_this) {
          return function(doneArgs) {
            return _this.trigger('itemRun', doneArgs);
          };
        })(this));
        item.on('success', (function(_this) {
          return function(doneArgs) {
            return _this.trigger('itemSuccess', doneArgs);
          };
        })(this));
        item.on('fail', (function(_this) {
          return function(failArgs, data) {
            return _this.trigger('itemFail', failArgs, data);
          };
        })(this));
        item.on('complete', (function(_this) {
          return function(doneOrFailArgs, data) {
            var wasTheLastItem;
            wasTheLastItem = _this._items.length === 1;
            _this._removeItem(item);
            _this.completeCount += 1;
            _this.trigger('itemComplete', doneOrFailArgs, data);
            if (wasTheLastItem) {
              return _this._handleTheLastItemCompletion();
            } else {
              return _this._handleNextRun();
            }
          };
        })(this));
        return this._items.push(item);
      };

      Pipeline.prototype.countRunningItems = function() {
        var item, n, _i, _len, _ref;
        n = 0;
        _ref = this._items;
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          item = _ref[_i];
          if (item.running) {
            n += 1;
          }
        }
        return n;
      };

      Pipeline.prototype.size = function() {
        return this._items.length;
      };

      Pipeline.prototype.run = function() {
        if (this.running) {
          return;
        }
        if (!this._items.length) {
          return;
        }
        this.running = true;
        this.trigger('run');
        return this._tryToRunNextItem();
      };

      Pipeline.prototype.stopAll = function() {
        var item, _i, _len, _ref;
        this._beforeStoppingItems();
        _ref = this._items;
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          item = _ref[_i];
          item.stop();
        }
        return this._afterStoppingItems();
      };

      Pipeline.prototype.stopAllWithoutTheLast = function() {
        var item, theLastItem, _i, _len, _ref;
        this._beforeStoppingItems();
        theLastItem = this._items[this._items.length - 1];
        _ref = this._items;
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          item = _ref[_i];
          if (item !== theLastItem) {
            item.stop();
          }
        }
        return this._afterStoppingItems();
      };

      return Pipeline;

    })(EveEve);
    $.DeferredPipelineNs = ns;
    return $.DeferredPipeline = ns.Pipeline;
  })(jQuery);

}).call(this);


jquery.deferredPipeLine.js
/*! jQuery.deferredPipeline (https://github.com/Takazudo/jQuery.deferredPipeline)
 * lastupdate: 2014-03-27
 * version: 0.0.0
 * author: 'Takazudo' Takeshi Takatsudo <takazudo@gmail.com>
 * License: MIT */
(function() {
  var __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  (function($) {
    var EveEve, ns, wait;
    EveEve = window.EveEve;
    ns = {};
    ns.util = {};
    wait = function(time) {
      return $.Deferred(function(defer) {
        return setTimeout(function() {
          return defer.resolve();
        }, time);
      });
    };
    ns.MESSAGE = {
      error_no_deferred: 'Error: registered function should return a deferred'
    };
    ns.util.isPromise = function(obj) {
      if ((obj != null) && (obj.then != null) && (obj.done != null)) {
        return true;
      }
      return false;
    };
    ns.Item = (function(_super) {
      __extends(Item, _super);

      Item.defaults = {
        done: null,
        fail: null,
        complete: null
      };

      function Item(fn, options) {
        this.options = $.extend({}, ns.Item.defaults, options);
        this.started = false;
        this.running = false;
        this.stopped = false;
        this._fn = fn;
      }

      Item.prototype._attachNoDeferredMessage = function(obj) {
        return obj.msg = ns.MESSAGE.error_no_deferred;
      };

      Item.prototype._triggerSuccess = function(doneArgs) {
        var _base;
        this.stopped = true;
        if (typeof (_base = this.options).done === "function") {
          _base.done(doneArgs);
        }
        this.trigger('success', doneArgs);
        return this._triggerComplete(doneArgs, true);
      };

      Item.prototype._triggerFail = function(failArgs, aborted, noDeferred) {
        var data, _base;
        if (aborted == null) {
          aborted = false;
        }
        if (noDeferred == null) {
          noDeferred = false;
        }
        this.stopped = true;
        data = {};
        data.aborted = aborted;
        if (noDeferred) {
          this._attachNoDeferredMessage(data);
        }
        if (typeof (_base = this.options).fail === "function") {
          _base.fail(failArgs, data);
        }
        this.trigger('fail', failArgs, data);
        return this._triggerComplete(failArgs, false, aborted, noDeferred);
      };

      Item.prototype._triggerComplete = function(doneOrFailArgs, successed, aborted, noDeferred) {
        var data, _base;
        if (aborted == null) {
          aborted = false;
        }
        if (noDeferred == null) {
          noDeferred = false;
        }
        data = {
          successed: successed,
          aborted: aborted
        };
        if (noDeferred) {
          this._attachNoDeferredMessage(data);
        }
        if (typeof (_base = this.options).complete === "function") {
          _base.complete(doneOrFailArgs, data);
        }
        return this.trigger('complete', doneOrFailArgs, data);
      };

      Item.prototype.destroy = function() {
        this.off();
        return this._fn = null;
      };

      Item.prototype.stop = function() {
        var _ref;
        if (this.stopped) {
          return;
        }
        this.stopped = true;
        if ((_ref = this._completeStats) != null) {
          _ref.aborted = true;
        }
        return this._triggerFail([], true);
      };

      Item.prototype.run = function() {
        var promise;
        this._completeStats = {
          aborted: false
        };
        this.started = true;
        if (this.stopped && (!this.running)) {
          return;
        }
        this.running = true;
        promise = this._fn(this._completeStats);
        if (!ns.util.isPromise(promise)) {
          this.running = false;
          this._triggerFail([], false, false, true);
          return;
        }
        this.trigger('run');
        return promise.then((function(_this) {
          return function() {
            _this.running = false;
            if (_this.stopped) {
              return;
            }
            return _this._triggerSuccess(arguments);
          };
        })(this), (function(_this) {
          return function() {
            _this.running = false;
            return _this._triggerFail(arguments, false);
          };
        })(this));
      };

      return Item;

    })(EveEve);
    ns.Pipeline = (function(_super) {
      __extends(Pipeline, _super);

      Pipeline.defaults = {
        pipeSize: 3
      };

      function Pipeline(options) {
        this.completeCount = 0;
        this.options = $.extend({}, ns.Pipeline.defaults, options);
        this._items = [];
      }

      Pipeline.prototype._beforeStoppingItems = function() {
        this._stoppingItemsInProgress = true;
        return this.trigger('startStoppingItems');
      };

      Pipeline.prototype._afterStoppingItems = function() {
        this._stoppingItemsInProgress = false;
        this.trigger('endStoppingItems');
        return this.trigger('stop');
      };

      Pipeline.prototype._handleTheLastItemCompletion = function() {
        this.running = false;
        return (wait(0)).done((function(_this) {
          return function() {
            return _this.trigger('allComplete');
          };
        })(this));
      };

      Pipeline.prototype._handleNextRun = function() {
        if (this._stoppingItemsInProgress) {
          return this.once('endStoppingItems', (function(_this) {
            return function() {
              return _this._tryToRunNextItem();
            };
          })(this));
        } else {
          return this._tryToRunNextItem();
        }
      };

      Pipeline.prototype._findNextPendingItem = function() {
        var item, _i, _len, _ref;
        _ref = this._items;
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          item = _ref[_i];
          if (item.started === false) {
            return item;
          }
        }
        return null;
      };

      Pipeline.prototype._removeItem = function(item) {
        var current, refreshed, _i, _len, _ref;
        refreshed = [];
        _ref = this._items;
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          current = _ref[_i];
          if (current !== item) {
            refreshed.push(current);
          }
        }
        this._items = refreshed;
        return item.destroy();
      };

      Pipeline.prototype._tryToRunNextItem = function() {
        var hitLimit, pipeSize, runOne, runningCount, _results;
        pipeSize = this.options.pipeSize;
        runningCount = this.countRunningItems();
        hitLimit = false;
        runOne = (function(_this) {
          return function() {
            var next;
            next = _this._findNextPendingItem();
            if (next != null) {
              next.run();
            } else {
              hitLimit = true;
            }
            return runningCount = _this.countRunningItems();
          };
        })(this);
        _results = [];
        while ((runningCount < pipeSize) && (!hitLimit)) {
          _results.push(runOne());
        }
        return _results;
      };

      Pipeline.prototype.destroy = function() {
        var item, _i, _len, _ref;
        this.stopAll();
        this.off();
        _ref = this._items;
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          item = _ref[_i];
          item.destroy();
        }
        return this._items = null;
      };

      Pipeline.prototype.add = function(fn, options) {
        var item;
        item = new ns.Item(fn, options);
        item.on('run', (function(_this) {
          return function(doneArgs) {
            return _this.trigger('itemRun', doneArgs);
          };
        })(this));
        item.on('success', (function(_this) {
          return function(doneArgs) {
            return _this.trigger('itemSuccess', doneArgs);
          };
        })(this));
        item.on('fail', (function(_this) {
          return function(failArgs, data) {
            return _this.trigger('itemFail', failArgs, data);
          };
        })(this));
        item.on('complete', (function(_this) {
          return function(doneOrFailArgs, data) {
            var wasTheLastItem;
            wasTheLastItem = _this._items.length === 1;
            _this._removeItem(item);
            _this.completeCount += 1;
            _this.trigger('itemComplete', doneOrFailArgs, data);
            if (wasTheLastItem) {
              return _this._handleTheLastItemCompletion();
            } else {
              return _this._handleNextRun();
            }
          };
        })(this));
        return this._items.push(item);
      };

      Pipeline.prototype.countRunningItems = function() {
        var item, n, _i, _len, _ref;
        n = 0;
        _ref = this._items;
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          item = _ref[_i];
          if (item.running) {
            n += 1;
          }
        }
        return n;
      };

      Pipeline.prototype.size = function() {
        return this._items.length;
      };

      Pipeline.prototype.run = function() {
        if (this.running) {
          return;
        }
        if (!this._items.length) {
          return;
        }
        this.running = true;
        this.trigger('run');
        return this._tryToRunNextItem();
      };

      Pipeline.prototype.stopAll = function() {
        var item, _i, _len, _ref;
        this._beforeStoppingItems();
        _ref = this._items;
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          item = _ref[_i];
          item.stop();
        }
        return this._afterStoppingItems();
      };

      Pipeline.prototype.stopAllWithoutTheLast = function() {
        var item, theLastItem, _i, _len, _ref;
        this._beforeStoppingItems();
        theLastItem = this._items[this._items.length - 1];
        _ref = this._items;
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          item = _ref[_i];
          if (item !== theLastItem) {
            item.stop();
          }
        }
        return this._afterStoppingItems();
      };

      return Pipeline;

    })(EveEve);
    $.DeferredPipelineNs = ns;
    return $.DeferredPipeline = ns.Pipeline;
  })(jQuery);

}).call(this);
